src/python/pants/backend/code_quality_tool/lib.py

```diff
@@ -0,0 +1,418 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from dataclasses import dataclass
+from typing import ClassVar, Iterable, Mapping
+
+from pants.core.goals.fmt import FmtFilesRequest, FmtResult
+from pants.core.goals.lint import LintFilesRequest, LintResult
+from pants.core.goals.run import RunFieldSet, RunInSandboxRequest
+from pants.core.util_rules.adhoc_process_support import (
+    ExtraSandboxContents,
+    MergeExtraSandboxContents,
+    ResolvedExecutionDependencies,
+    ResolveExecutionDependenciesRequest,
+)
+from pants.core.util_rules.adhoc_process_support import rules as adhoc_process_support_rules
+from pants.core.util_rules.environments import EnvironmentNameRequest
+from pants.core.util_rules.partitions import Partitions
+from pants.engine.addresses import Addresses
+from pants.engine.environment import EnvironmentName
+from pants.engine.fs import PathGlobs
+from pants.engine.internals.native_engine import (
+    EMPTY_DIGEST,
+    Address,
+    AddressInput,
+    Digest,
+    FilespecMatcher,
+    MergeDigests,
+    Snapshot,
+)
+from pants.engine.internals.selectors import Get
+from pants.engine.process import FallibleProcessResult, Process
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    SpecialCasedDependencies,
+    StringField,
+    StringSequenceField,
+    Target,
+    Targets,
+)
+from pants.option.option_types import SkipOption
+from pants.option.subsystem import Subsystem
+from pants.util.frozendict import FrozenDict
+from pants.util.strutil import Simplifier, help_text
+
+
+class CodeQualityToolFileGlobIncludeField(StringSequenceField):
+    alias: ClassVar[str] = "file_glob_include"
+    required = True
+
+
+class CodeQualityToolFileGlobExcludeField(StringSequenceField):
+    alias: ClassVar[str] = "file_glob_exclude"
+    required = False
+    default = ()
+
+
+class CodeQualityToolRunnableField(StringField):
```

Comment:
If/when we move this to be more closely tied to the `adhoc` backend, it might be good to have these fields (including `args`, `execution_dependencies` etc. all inherit from the same class, similar to how `ShellCommandExecutionDependenciesField` inherits from `AdhocToolExecutionDependenciesField`.